### PR TITLE
Added Preview Builds via Gitlab CI

### PR DESCRIPTION
**Pull request by AlexGilleran**
### What this PR does
- Adds a complete build on Gitlab CI (this should be able to replace travis and circleci in time).
- Adds the ability to run previews on Dev - click the Gitlab link down the bottom when the build is finished and there'll be a "Run As Preview" link - this will deploy the branch to http://<branch-name>.magda-dev.terria.io in its own namespace in the dev cluster, and pull the latest backups for databases and elasticsearch from dev so the data should be reasonable for testing. It _won't_ run any sleuthers unless you use helm to add them manually.
- Changes the build scripts a bit to support this
- Makes changes to the way that backups work for the databases and search in order to support spinning up a new namespace and quickly loading data in - particularly giving databases the ability to only replay WAL logs until they're consistent, rather than replaying everything.
- Adds a manual snapshot feature to indexer.

### Checklist
- [x] Unit tests aren't applicable (delete one)
- [x] I've updated CHANGES.md with what I changed.
- [x] I've linked this PR to an issue in ZenHub (or there is no issue) and dragged it to the _Pull Request_ column

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -50,7 +50,14 @@ matrix:
         # Cleanup the cached directories to avoid unnecessary cache updates
         - find $HOME/.ivy2/cache -name "ivydata-*.properties" -print -delete
         - find $HOME/.sbt        -name "*.lock"               -print -delete
-      script: npm install -g lerna && npm run node-ci
+      script: 
+        - npm install -g lerna
+        - npm run in-submodules -- -f categories.api=true -f language=typescript -- bootstrap --include-filtered-dependencies
```
Comment (t83714) [inline .travis.yml]: Should we also run `build` in each of the modules? So any possible errors during build process will be discovered?

Comment (AlexGilleran) [reply .travis.yml]: Yeah good point, done :).

**Comment by t83714** [comment]
Tried to build and push docker images to my local Minikube (using `lerna run docker-build-local --include-filtered-dependencies`):
- first time had an error with `magda-builder-docker` said manifest for `magda-builder-nodejs`. I guess it happened because lerna picked `magda-builder-docker` to run before `magda-builder-nodejs` pushed to the registry.  Manually, push `magda-builder-nodejs and then tried `magda-builder-docker`. Worked this time.
- Wanted to double check this. Therefore, I re-deployed an empty registry in Minikube. And then lerna run docker-build-local --include-filtered-dependencies` again. This time and got different error:
```
> @magda/builder-docker@0.0.38-0 docker-build-local /Users/jia047/development/magda4/magda-builder-docker
> create-docker-context-for-node-component --build --push --tag auto --local

Sending build context to Docker daemon  9.068MB
Step 1/10 : FROM localhost:5000/data61/magda-builder-nodejs:latest
 ---> 6ab340cb50b2
Step 2/10 : RUN apk add --no-cache              ca-certificates
 ---> Using cache
 ---> e2a34d4df0a3
Step 3/10 : RUN [ ! -e /etc/nsswitch.conf ] && echo 'hosts: files dns' > /etc/nsswitch.conf
 ---> Using cache
 ---> f27f3ba8f492
Step 4/10 : ENV DOCKER_CHANNEL stable
 ---> Using cache
 ---> 8a8f2a297e21
Step 5/10 : ENV DOCKER_VERSION 18.03.0-ce
 ---> Using cache
 ---> 9698851e65b9
Step 6/10 : RUN set -ex;        apk add --no-cache --virtual .fetch-deps                curl            tar     ;               apkArch="$(apk --print-arch)";    case "$apkArch" in              x86_64) dockerArch='x86_64' ;;          armhf) dockerArch='armel' ;;            aarch64) dockerArch='aarch64' ;;          ppc64le) dockerArch='ppc64le' ;;                s390x) dockerArch='s390x' ;;            *) echo >&2 "error: unsupported architecture ($apkArch)"; exit 1 ;;       esac;           if ! curl -fL -o docker.tgz "https://download.docker.com/linux/static/${DOCKER_CHANNEL}/${dockerArch}/docker-${DOCKER_VERSION}.tgz"; then                 echo >&2 "error: failed to download 'docker-${DOCKER_VERSION}' from '${DOCKER_CHANNEL}' for '${dockerArch}'";             exit 1;         fi;             tar --extract           --file docker.tgz                 --strip-components 1            --directory /usr/local/bin/     ;       rm docker.tgz;          apk del .fetch-deps;              dockerd -v;     docker -v
 ---> Using cache
 ---> 841b6a973695
Step 7/10 : COPY component/modprobe.sh /usr/local/bin/modprobe
 ---> Using cache
 ---> e61763a841e2
Step 8/10 : COPY component/docker-entrypoint.sh /usr/local/bin/
 ---> Using cache
 ---> cdcf7a81d072
Step 9/10 : ENTRYPOINT docker-entrypoint.sh
 ---> Using cache
 ---> a623668d9868
Step 10/10 : CMD sh
 ---> Using cache
 ---> c6beff264db3
[WARNING]: Empty continuation line found in:
    RUN set -ex;        apk add --no-cache --virtual .fetch-deps                curl            tar     ;               apkArch="$(apk --print-arch)";    case "$apkArch" in              x86_64) dockerArch='x86_64' ;;          armhf) dockerArch='armel' ;;            aarch64) dockerArch='aarch64' ;;          ppc64le) dockerArch='ppc64le' ;;                s390x) dockerArch='s390x' ;;            *) echo >&2 "error: unsupported architecture ($apkArch)"; exit 1 ;;       esac;           if ! curl -fL -o docker.tgz "https://download.docker.com/linux/static/${DOCKER_CHANNEL}/${dockerArch}/docker-${DOCKER_VERSION}.tgz"; then                 echo >&2 "error: failed to download 'docker-${DOCKER_VERSION}' from '${DOCKER_CHANNEL}' for '${dockerArch}'";             exit 1;         fi;             tar --extract           --file docker.tgz                 --strip-components 1            --directory /usr/local/bin/     ;       rm docker.tgz;          apk del .fetch-deps;              dockerd -v;     docker -v
[WARNING]: Empty continuation lines will become errors in a future release.
Successfully built c6beff264db3
Successfully tagged localhost:5000/data61/magda-builder-docker:latest
The push refers to a repository [localhost:5000/data61/magda-builder-docker]

    at Promise.all.then.arr (/usr/local/lib/node_modules/lerna/node_modules/execa/index.js:236:11)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:160:7)
  code: 1,
  killed: false,
```
Is that my local issue or we really need to remove those empty lines?

**Comment by AlexGilleran** [comment]
I removed those empty lines.
For the builders, you need to build nodejs then docker then scala. If you use
```
npm run in-submodules -- -f categories.builder=true -- run docker-build-local --include-filtered-dependencies
```
They'll all build in order. You can also see them being built in the first stage of the pipeline.

**Pull request merged**